### PR TITLE
Image Preprocessor

### PR DESCRIPTION
**Pull request by gabrielpoca**
This change introduces the **optional** image preprocessor built on top of [`imageflow_ex`](https://github.com/naps62/imageflow_ex) which defaults to ["mogrify"](https://github.com/route/mogrify) which uses ImageMagick.
The API is still low-level, but it's a start. 

The preprocessor can generate multiple output files and apply transformations to them. 

Image processing is pretty slow, so we use the files's modified time and name to decide whether to run or not: if the output files already exist and the input file's modified time is before, or the same, as the output files', it doesn't run.

The transformations map is hashed and encoded in the file's name, so we know that if we change the transformation, the expected output name will also change, and the preprocessor will run.

Besides the preprocessor, there's also a `ResponsiveImage` view helper that sets the sizes automatically based on the input file's width. It's a pretty dump setup for now.

Extra:

* `variables` was renamed to `metadata` because it was starting to get confusing. I find that metadata is a better name.
* the compilation timeout was increased by 5 seconds before processing images is a slow operation and sometimes it times out. Keep in mind that this operation is only slow when we are building the site for the first time or when we remove the output folder.

### CODE REVIEW

**Comment by gabrielpoca** [comment]
cc @naps62

File: lib/still/compiler/view_helpers/responsive_image.ex [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  require Logger
+
+  @default_nr_of_sizes 4
+
+  @doc """
+  Returns an image tag with the `src` and `srcset`.
+
+  If the `:image_opts` is set in `opts`, it will be passed on to #{Preprocessor.Image}.
+
+  If the `:sizes` is not set in the `:image_opts` map, it sets the sizes as as
```
Comment (frm) [inline lib/still/compiler/view_helpers/responsive_image.ex:17]: > Drop the `the`. It's cleaner.

-- Justin Timberlake

File: lib/still/compiler/view_helpers/responsive_image.ex [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  import Still.Utils
+
+  require Logger
+
+  @default_nr_of_sizes 4
+
+  @doc """
+  Returns an image tag with the `src` and `srcset`.
+
+  If the `:image_opts` is set in `opts`, it will be passed on to #{Preprocessor.Image}.
```
Comment (frm) [inline lib/still/compiler/view_helpers/responsive_image.ex:15]: > Drop the `the`. It's cleaner.

-- Justin Timberlake

File: lib/still/compiler/view_helpers/responsive_image.ex [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    output_files
+    |> Enum.map(fn {size, file} ->
+      "#{UrlFor.render(file)} #{size}w"
+    end)
+    |> Enum.join(", ")
+  end
+
+  @spec get_render_data(String.t(), any()) :: %{
+          image_opts: Preprocessor.ResponsiveImage.opts()
+        }
```
Comment (frm) [inline lib/still/compiler/view_helpers/responsive_image.ex:59]: Why do we have this spec for a private function?

File: lib/still/preprocessor.ex [hunk 4]
```diff
@@ -16,25 +16,28 @@ defmodule Still.Preprocessor do
...
+    Image
   }
 
   @default_preprocessors %{
-    ".slim" => [AddContent, Frontmatter, Slime, OutputPath, AddLayout, Save],
-    ".slime" => [AddContent, Frontmatter, Slime, OutputPath, AddLayout, Save],
-    ".eex" => [AddContent, Frontmatter, EEx, OutputPath, AddLayout, Save],
+    ".slim" => [AddContent, EEx, Frontmatter, Slime, OutputPath, AddLayout, Save],
+    ".slime" => [AddContent, EEx, Frontmatter, Slime, OutputPath, AddLayout, Save],
+    ".eex" => [AddContent, EEx, Frontmatter, EEx, OutputPath, AddLayout, Save],
```
Comment (frm) [inline lib/still/preprocessor.ex:26]: Is this intentional? You have `EEx` twice

File: lib/still/preprocessor/image.ex [hunk 5]
```diff
@@ -0,0 +1,139 @@
+defmodule Still.Preprocessor.Image do
+  @moduledoc """
+  This preprocessor handles image transormation. To configure
```
Comment (frm) [inline lib/still/preprocessor/image.ex:3]: transformation

File: lib/still/preprocessor/image.ex [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  it, set the `:image_opts` in the metadata:
```
Comment (frm) [inline lib/still/preprocessor/image.ex:4]: > Drop the `the`. It's cleaner.

-- Justin Timberlake

File: lib/still/preprocessor/image.ex [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+    %{
+      image_opts: %{
+        sizes: [100, 200],
+        transformations: [color_filter: "grayscale_bt709"]
+      }
+    }
+
+  The `:sizes` defines the widths of the output files to create.
```
Comment (frm) [inline lib/still/preprocessor/image.ex:13]: > Drop the `the`. It's cleaner.

-- Justin Timberlake

(or append `options`: "The `:sizes` option)

File: lib/still/preprocessor/image.ex [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+  The `:transformations` defines the function name and arguments to call on the
```
Comment (frm) [inline lib/still/preprocessor/image.ex:15]: > Drop the `the`. It's cleaner.

-- Justin Timberlake

File: test/support/still/case.ex [hunk 9]
```diff
@@ -9,6 +9,7 @@ defmodule Still.Case do
         Application.put_env(:still, :pass_through_copy, [])
 
         Still.Utils.rm_output_dir()
+        Still.Utils.mk_output_dir()
```
Comment (frm) [inline test/support/still/case.ex:12]: Why do we need this?

**Review by frm — CHANGES_REQUESTED** [review]

File: .github/workflows/elixir.yml [hunk 10]
```diff
@@ -11,6 +11,15 @@ jobs:
     name: Build and test
     runs-on: ubuntu-latest
 
+    strategy:
+      matrix:
+        otp:
+          - 23
+        elixir:
+          - 1.10.4
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply .github/workflows/elixir.yml:19]: Maybe? I don't think we are using anything special that wasn't already available in previous versions.

File: lib/still/compiler/view_helpers.ex [hunk 11]
```diff
@@ -60,11 +78,19 @@ defmodule Still.Compiler.ViewHelpers do
...
 
       def link_to_css(path, opts \\ []) do
-        LinkToCSS.render(path, opts, @env)
+        LinkToCSS.render(path, opts)
       end
 
       def link_to_js(path, opts \\ []) do
-        LinkToJS.render(path, opts, @env)
+        LinkToJS.render(path, opts)
+      end
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply lib/still/compiler/view_helpers.ex:86]: 👍

File: lib/still/compiler/view_helpers/content_tag.ex [hunk 12]
```diff
@@ -10,7 +10,7 @@ defmodule Still.Compiler.ViewHelpers.ContentTag do
       |> Enum.concat(data_attrs)
       |> Enum.join(" ")
 
-    opening_tag(tag, content, attrs) <> content <> closing_tag(tag, content)
+    opening_tag(tag, content, attrs) <> (content || "") <> closing_tag(tag, content)
```
[the remark this answers is not in the document]
Comment (gabrielpoca) [reply lib/still/compiler/view_helpers/content_tag.ex:13]: The `closing_tag` function already takes care of that. But I don't think that's true, if I'm not mistaken, not all elements are valid self-closing tags.

File: lib/still/compiler/view_helpers/responsive_image.ex (hunk 3 above)
Comment (gabrielpoca) [reply lib/still/compiler/view_helpers/responsive_image.ex:59]: 👍 it wasn't private at some point

File: lib/still/preprocessor.ex (hunk 4 above)
Comment (gabrielpoca) [reply lib/still/preprocessor.ex:26]: No 👍

File: test/support/still/case.ex (hunk 9 above)
Comment (gabrielpoca) [reply test/support/still/case.ex:12]: For the output folder to exist

Comment (frm) [reply test/support/still/case.ex:12]: Should we not remove it in the line above, then? Just remove the contents

Comment (gabrielpoca) [reply test/support/still/case.ex:12]: No, because there are files we want to remove there.

Comment (frm) [reply test/support/still/case.ex:12]: Yeah, but can't we remove the contents in a single system call? `Still.Utils.clean_output_dir/0` that does `File.rm(dir <> "/*")`

File: lib/still/compiler/view_helpers.ex [hunk 13]
```diff
@@ -77,13 +77,9 @@ defmodule Still.Compiler.ViewHelpers do
...
 
-      def link_to_css(path, opts \\ []) do
-        LinkToCSS.render(path, opts)
-      end
+      defdelegate link_to_css(path, opts \\ []), to: LinkToCSS, as: :render
 
-      def link_to_js(path, opts \\ []) do
-        LinkToJS.render(path, opts)
-      end
+      defdelegate link_to_js(path, opts \\ []), to: LinkToJS, as: :render
```
Comment (frm) [inline lib/still/compiler/view_helpers.ex:82]: 👍

File: test/support/still/case.ex (hunk 9 above)
Comment (gabrielpoca) [reply test/support/still/case.ex]: I'm not sure if having one system call instead of those two is better, but sure 👍

File: lib/still/preprocessor/image.ex [hunk 14]
```diff
@@ -23,40 +23,28 @@ defmodule Still.Preprocessor.Image do
...
-
-    if input_file_changed?(input_file, output_files) do
-      process_input_file(input_file, opts, output_files)
-    end
-
-    %{
-      source_file
-      | metadata: Map.put(metadata, :image_output_files, output_files)
-    }
+  case Code.ensure_compiled(Imageflow) do
```
Comment (frm) [inline lib/still/preprocessor/image.ex:28]: Use `Code.ensure_loaded?/1` instead.

File: lib/still/preprocessor/renderer.ex [hunk 15]
```diff
@@ -29,7 +29,7 @@ defmodule Still.Preprocessor.Renderer do
           |> Enum.map(&String.replace(&1, "_", ""))
           |> Enum.map(&String.capitalize/1)
 
-        Module.concat([@preprocessor | name])
+        Module.concat(["R#{Enum.random(0..100_000)}" | [@preprocessor | name]])
```
Comment (frm) [inline lib/still/preprocessor/renderer.ex:32]: What's the reason for this change?

File: lib/still/utils.ex [hunk 16]
```diff
@@ -24,14 +23,37 @@ defmodule Still.Utils do
...
 
-  def get_image_info(file) do
-    job = Native.create!()
-
-    with :ok <- Native.add_input_file(job, 0, file),
-         {:ok, %{"code" => 200, "data" => %{"image_info" => image_info}}} <-
-           Native.message(job, "v0.1/get_image_info", %{io_id: 0}) do
-      image_info
-    end
+  case Code.ensure_compiled(Imageflow) do
```
Comment (frm) [inline lib/still/utils.ex:26]: Use `Conde.ensure_loaded?/1` instead.

File: test/still/preprocessor/image_test.exs [hunk 17]
```diff
@@ -124,7 +124,7 @@ defmodule Still.Preprocessor.ImageTest do
           metadata: %{
             image_opts: %{
               sizes: [100, 200],
-              transformations: [color_filter: "grayscale_bt709"]
```
Comment (frm) [inline test/still/preprocessor/image_test.exs]: Is this removal intended? Should we replace it with another one?

**Review by frm — CHANGES_REQUESTED** [review]

File: test/still/preprocessor/image_test.exs (hunk 17 above)
Comment (gabrielpoca) [reply test/still/preprocessor/image_test.exs]: It is because otherwise, the tests fail when changing between imageflow and imagemagick.

File: lib/still/preprocessor/renderer.ex (hunk 15 above)
Comment (frm) [reply lib/still/preprocessor/renderer.ex:32]: @gabrielpoca

Comment (frm) [reply lib/still/preprocessor/renderer.ex:32]: Via call we concluded this was the same problem redefining modules I was facing before. This is a fair approach for now 👍

**Review by frm — APPROVED** [review]
